cub/agent/agent_batch_memcpy.cuh

```diff
@@ -0,0 +1,967 @@
+  data_out.w = __funnelshift_r(data_out.w, tail, bit_shift);
+}
+
+__forceinline__ __device__ void LoadVectorAndFunnelShiftR(uint32_t const *aligned_ptr,
+                                                          uint32_t bit_shift,
+                                                          uint2 &data_out)
+{
+  data_out      = {aligned_ptr[0], aligned_ptr[1]};
+  uint32_t tail = 0;
+  if (bit_shift)
+    tail = aligned_ptr[2];
+
+  data_out.x = __funnelshift_r(data_out.x, data_out.y, bit_shift);
+  data_out.y = __funnelshift_r(data_out.y, tail, bit_shift);
+}
+
+__forceinline__ __device__ void LoadVectorAndFunnelShiftR(uint32_t const *aligned_ptr,
+                                                          uint32_t bit_shift,
+                                                          uint32_t &data_out)
+{
+  data_out      = aligned_ptr[0];
+  uint32_t tail = 0;
+  if (bit_shift)
+    tail = aligned_ptr[1];
+
+  data_out = __funnelshift_r(data_out, tail, bit_shift);
+}
+
+template <typename VectorT>
+__forceinline__ __device__ void LoadVector(const char *ptr, VectorT &data_out)
+{
+  const uint32_t offset            = reinterpret_cast<std::uintptr_t>(ptr) % 4U;
+  const uint32_t *aligned_ptr      = reinterpret_cast<uint32_t const *>(ptr - offset);
+  constexpr uint32_t bits_per_byte = 8U;
+  const uint32_t bit_shift         = offset * bits_per_byte;
+
+  LoadVectorAndFunnelShiftR(aligned_ptr, bit_shift, data_out);
+}
+
+/**
+ * @brief Cooperatively copies \p num_bytes from \p src to \p dest using vectorized stores of type \p VectorT for
+ * addresses within [dest, dest + num_bytes) that are aligned to \p VectorT. A byte-wise copy is used for byte-ranges
+ * that are not aligned to \p VectorT.
+ *
+ * @tparam VectorT The vector type used for vectorized stores (i.e., one of uint4, uint2, uint32_t)
+ * @tparam ByteOffsetT Type sufficiently large to
+ * @param thread_rank The thread rank within the group that cooperates to copy the data must be within [0, `group_size`)
+ * @param group_size The number of threads cooperaing to copy the data; all threads within [0, `group_size`) must invoke
+ * this method with the same arguments
+ * @param dest Pointer to the memory location to copy to
+ * @param num_bytes Number of bytes to copy
+ * @param src Pointer to the memory location to copy from
+ */
+template <typename VectorT, typename ByteOffsetT>
+__device__ __forceinline__ void
+VectorizedCopy(int32_t thread_rank, int32_t group_size, void *dest, ByteOffsetT num_bytes, const void *src)
+{
+  char *out_ptr                      = reinterpret_cast<char *>(dest);
+  const char *in_ptr                 = reinterpret_cast<const char *>(src);
+  constexpr size_t out_datatype_size = sizeof(VectorT);
+  constexpr size_t in_datatype_size  = sizeof(uint32_t);
+
+  // Number of bytes between out_ptr and the first OutT-aligned address at or after out_ptr
+  auto const alignment_offset = reinterpret_cast<std::uintptr_t>(out_ptr) % out_datatype_size;
```

Comment:
I believe the code base favors `const auto` over `auto const`

Also should we have a macro for that?